api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -723,6 +723,9 @@ public int hashCode() {
 
   public static class StructType extends NestedType {
     private static final Joiner FIELD_SEP = Joiner.on(", ");
+    private static final int NO_HASHCODE = Integer.MIN_VALUE;
+
+    private transient int hashCode = NO_HASHCODE;
```

Comment:
Thanks for making this transient, when I saw this PR that's the one aspect I wanted to make sure was the case. We really don't want any weird issues that will happen in a distributed setting where the cached hashcode on one struct type is different than the hashcode for the same struct type that's on a different node (which can easily happen since hashcodes across JVMS can be different)

Making it transient will avoid all those kinds of issues.